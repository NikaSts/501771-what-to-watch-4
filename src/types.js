import {string, number, shape, arrayOf, bool} from 'prop-types';

export const movieType = shape({
  id: number.isRequired,
  title: string.isRequired,
  runTime: number.isRequired,
  genre: string.isRequired,
  releaseDate: number.isRequired,
  poster: string.isRequired,
  previewImage: string.isRequired,
  backgroundImage: string.isRequired,
  backgroundColor: string.isRequired,
  ratingScore: number.isRequired,
  ratingLevel: string.isRequired,
  ratingCount: number.isRequired,
  paragraphs: arrayOf(string).isRequired,
  director: string.isRequired,
  stars: arrayOf(string).isRequired,
  preview: string.isRequired,
  isFavorite: bool.isRequired,
});

export const promoMovieType = shape({
  id: number.isRequired,
  title: string.isRequired,
  genre: string.isRequired,
  releaseDate: number.isRequired,
  poster: string.isRequired,
  backgroundImage: string.isRequired,
  backgroundColor: string.isRequired,
});

export const cardMovieType = shape({
  title: string.isRequired,
  image: string.isRequired,
});
