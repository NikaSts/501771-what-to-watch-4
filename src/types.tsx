export interface MovieType {
  id: number;
  title: string;
  runTime: number;
  genre: string;
  releaseDate: number;
  poster: string;
  previewImage: string;
  backgroundImage: string;
  backgroundColor: string;
  ratingScore: number;
  ratingLevel?: string;
  ratingCount: number;
  description: string;
  director: string;
  stars: Array<string>;
  preview: string;
  video: string;
  isFavorite: boolean;
}

export interface UserDataType {
  id: number;
  name: string;
  email: string;
  avatar: string;
}

export interface ReviewType {
  id: number;
  user: {
    id: number;
    name: string;
  };
  rating: number;
  comment: string;
  date: string;
}
