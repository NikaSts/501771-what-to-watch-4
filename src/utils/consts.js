export const Tab = {
  OVERVIEW: `Overview`,
  REVIEWS: `Reviews`,
  DETAILS: `Details`,
};

export const tabs = [Tab.OVERVIEW, Tab.REVIEWS, Tab.DETAILS];

export const DEFAULT_GENRE = `All genres`;

export const MOVIES_TO_SHOW_COUNT = 8;
export const MAX_GENRES_TO_SHOW = 10;
export const MAX_SIMILAR_MOVIES = 4;
export const MAX_ACTORS_TO_SHOW = 4;

export const Rating = {
  BAD: `Bad`,
  NORMAL: `Normal`,
  GOOD: `Good`,
  VERY_GOOD: `Very good`,
  AWESOME: `Awesome`,
};

export const Score = {
  BAD: 60,
  NORMAL: 70,
  GOOD: 80,
  VERY_GOOD: 90,
  AWESOME: 100,
};

export const DELAY = 1000;

export const ActionType = {
  OPEN_FULL_SCREEN_PLAYER: `openFullScreenPlayer`,
  CLOSE_FULL_SCREEN_PLAYER: `closeFullScreenPlayer`,
  GET_MOVIES: `getMovies`,
  GET_PROMO_MOVIE: `getPromoMovie`,
  GET_REVIEWS: `getReviews`,
  REQUIRE_AUTHORIZATION: `requireAuthorization`,
  IS_AUTHORIZING: `isAuthorizing`,
  IS_NOT_AUTHORIZING: `isNotAuthorizing`,
  IS_AUTHORIZATION_ERROR: `isAuthorizationError`,
  LOAD_USER_DATA: `loadUserData`,
};

export const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

export const EntryPoint = {
  REVIEWS: `/wtw/comments`,
  FAVORITE: `/wtw/favorite`,
  MOVIES: `/wtw/films`,
  PROMO: `/wtw/films/promo`,
  LOGIN: `/wtw/login`,
};

export const TIMER = 5000;

export const Error = {
  UNAUTHORIZED: 401,
};

export const URL = `https://4.react.pages.academy`;

export const AppRoute = {
  ROOT: `/`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  MOVIE_PAGE: `/films`,
};

export const TEXTAREA_COLOR = `rgba(255, 255, 255, 0.3)`;
