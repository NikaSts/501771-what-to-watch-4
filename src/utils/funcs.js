import {DEFAULT_GENRE, Rating, Score} from './consts';

export const getRandomNumber = (min, max) => {
  return Math.floor(Math.random() * (max - min)) + min;
};

export const getRandomItem = (items) => {
  return items[getRandomNumber(0, items.length)];
};

export const getUrlFromTitle = (title) => {
  return title
    .toLowerCase()
    .split(` `)
    .join(`-`);
};

export const getRandomBoolean = () => {
  return Math.random() > 0.5;
};

export const getRandomItems = (items) => {
  const result = [];
  [...items].forEach((item) => {
    if (getRandomBoolean()) {
      result.push(item);
    }
  });
  return result;
};

export const suffleItems = (items) => {
  for (let i = items.length - 1; i > 0; i--) {
    const randomIndex = Math.floor(Math.random() * (i + 1));
    const temp = items[randomIndex];
    items[randomIndex] = items[i];
    items[i] = temp;
  }
  return items;
};

export const getRatingLevel = (score) => {
  if (score < Score.BAD) {
    return Rating.BAD;
  }
  if (score < Score.NORMAL) {
    return Rating.NORMAL;
  }
  if (score < Score.GOOD) {
    return Rating.GOOD;
  }
  if (score < Score.VERY_GOOD) {
    return Rating.VERY_GOOD;
  }
  return Rating.AWESOME;
};

export const getMoviesToShow = (movies, activeGenre) => {
  if (activeGenre === DEFAULT_GENRE) {
    return [...movies];
  }
  return [...movies].filter((movie) => movie.genre === activeGenre);
};
