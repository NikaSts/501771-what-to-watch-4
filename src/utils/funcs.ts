
import {DEFAULT_GENRE, Rating, Score} from './consts';
import {MovieType} from '../types';

interface ExtendInterface {
    <T>(state: T, newStateValue: T): T;
}

export const extend: ExtendInterface = (state, newStateValue) => {
  return Object.assign({}, state, newStateValue);
};

export const getRatingLevel = (score: number) => {
  if (score < Score.BAD) {
    return Rating.BAD;
  }
  if (score < Score.NORMAL) {
    return Rating.NORMAL;
  }
  if (score < Score.GOOD) {
    return Rating.GOOD;
  }
  if (score < Score.VERY_GOOD) {
    return Rating.VERY_GOOD;
  }
  return Rating.AWESOME;
};

export const filterMovies = (movies: Array<MovieType>, activeGenre: string = DEFAULT_GENRE) => {
  if (activeGenre === DEFAULT_GENRE) {
    return [...movies];
  }
  return [...movies].filter((movie) => movie.genre === activeGenre);
};

export const getGenresFromMovies = (movies: Array<MovieType>) => {
  return [DEFAULT_GENRE, ...new Set(movies.map((movie) => movie.genre))];
};

export const getMovieById = (movies: Array<MovieType>, id: number) => {
  return movies.find((movie) => movie.id === id);
};

export const formatTimeWithSeconds = (time: number) => {
  const minutes = Math.trunc(time / 60);
  const seconds = Math.trunc(time % 60);
  const hours = Math.trunc(minutes / 60);

  return [
    (`0` + hours).slice(-2),
    (`0` + minutes).slice(-2),
    (`0` + seconds).slice(-2)
  ].join(`:`);
};

export const formatTime = (time: number) => {
  const hours = Math.trunc(time / 60);
  const minutes = time % 60;

  if (hours) {
    return `${hours}h ${minutes}m`;
  }
  return `${minutes}m`;
};
