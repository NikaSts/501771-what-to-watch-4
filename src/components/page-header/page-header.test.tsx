import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Router} from 'react-router-dom';
import history from '../../history';
import PageHeader from './page-header';
import {Page} from '../../utils/consts';
import {UserDataType} from '../../types';

const userData: UserDataType = {
  id: 1,
  name: ``,
  email: ``,
  avatar: ``,
};
const children: React.ReactNode = <div></div>;

it(`PageHeader on MainPage, not authtorized should render correctly`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <PageHeader
            currentPage={Page.MAIN}
            isAuth={false}
            user={userData}>
            {children}
          </PageHeader>
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`PageHeader on MainPage, authtorized should render correctly`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <PageHeader
            currentPage={Page.MAIN}
            isAuth={true}
            user={userData}>
            {children}
          </PageHeader>
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});


it(`PageHeader on LoginPage, not authtorized should render correctly`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <PageHeader
            currentPage={Page.SIGN_IN}
            isAuth={false}
            user={userData}>
            {children}
          </PageHeader>
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`PageHeader on MyListPage should render correctly`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <PageHeader
            currentPage={Page.MY_LIST}
            isAuth={true}
            user={userData}>
            {children}
          </PageHeader>
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`PageHeader on ReviewPage should render correctly`, () => {
  const tree = renderer
    .create(
        <Router history={history}>
          <PageHeader
            currentPage={Page.REVIEW}
            isAuth={true}
            user={userData}>
            {children}
          </PageHeader>
        </Router>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
