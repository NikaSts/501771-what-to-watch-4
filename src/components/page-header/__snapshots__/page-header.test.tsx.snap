// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageHeader on LoginPage, not authtorized should render correctly 1`] = `
<header
  className="page-header user-page__head"
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
      href="/"
      onClick={[Function]}
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <h1
    className="page-title user-page__title"
  >
    Sign in
  </h1>
</header>
`;

exports[`PageHeader on MainPage, authtorized should render correctly 1`] = `
<header
  className="page-header movie-card__head"
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
      href="/"
      onClick={[Function]}
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <div
    className="user-block"
  >
    <a
      href="/mylist"
      onClick={[Function]}
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="https://4.react.pages.academy"
          width="63"
        />
      </div>
    </a>
  </div>
</header>
`;

exports[`PageHeader on MainPage, not authtorized should render correctly 1`] = `
<header
  className="page-header movie-card__head"
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
      href="/"
      onClick={[Function]}
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <div
    className="user-block"
  >
    <a
      className="user-block__link"
      href="/login"
      onClick={[Function]}
    >
      Sign in
    </a>
  </div>
</header>
`;

exports[`PageHeader on MyListPage should render correctly 1`] = `
<header
  className="page-header user-page__head"
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
      href="/"
      onClick={[Function]}
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <h1
    className="page-title user-page__title"
  >
    My list
  </h1>
  <div
    className="user-block"
  >
    <a
      href="/mylist"
      onClick={[Function]}
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="https://4.react.pages.academy"
          width="63"
        />
      </div>
    </a>
  </div>
</header>
`;

exports[`PageHeader on ReviewPage should render correctly 1`] = `
<header
  className="page-header movie-card__head"
>
  <div
    className="logo"
  >
    <a
      className="logo__link"
      href="/"
      onClick={[Function]}
    >
      <span
        className="logo__letter logo__letter--1"
      >
        W
      </span>
      <span
        className="logo__letter logo__letter--2"
      >
        T
      </span>
      <span
        className="logo__letter logo__letter--3"
      >
        W
      </span>
    </a>
  </div>
  <div />
  <div
    className="user-block"
  >
    <a
      href="/mylist"
      onClick={[Function]}
    >
      <div
        className="user-block__avatar"
      >
        <img
          alt="User avatar"
          height="63"
          src="https://4.react.pages.academy"
          width="63"
        />
      </div>
    </a>
  </div>
</header>
`;
